SCALRCORE-16917: Use Revizor API to create test containers to run acceptance tests against

vmotso opened the pull request.

vmotso committed:
Init revizor GitHub Action

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Add Dockerfile

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Add skip_ui to speed up the container creation

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Update log messages

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Add action.yml

Action metadata file
---
Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Update Dockerfile

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Check if required environment variables were set

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Use secret for revizor url

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Add output to the action

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Add hostname to outputs

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Fix set-output

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Add the CI pipeline to run lint

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Update log messages

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Check if env variable is not empty

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Decrease the revizor API call timeout

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Setup revizor container branches

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Add README.md

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso commented:
All tests were successful:
https://github.com/vmotso/check-gh-action-revizor/runs/1510664320

To see how to run tests with the container from a specific branch
https://github.com/vmotso/check-gh-action-revizor/actions/runs/405820559

P.S.
I added access to the test repository
https://github.com/vmotso/check-gh-action-revizor

You can see Secrets and edit it.

lbn commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+COPY main.go .
+RUN CGO_ENABLED=0 GOOS=linux go build -o revizor .
+
+FROM golang:1.15
```
It's a 839MB image. Can you try `scratch` instead?

lbn commented on README.md:
```diff
@@ -1 +1,35 @@
-# gh-action-revizor
+# revizor docker action
+
+This action `create|delete` the test environemnt container.
```
Typo

lbn commented on README.md:
```diff
@@ -1 +1,35 @@
+
+**Required** The command to execute: `create|delete`.
+
+### `conatiner_id`
```
Typo (same in output).

lbn commented on main.go:
```diff
@@ -0,0 +1,182 @@
+	// Such a large timeout due to the fact that sometimes
+	// the result of creating a container cannot be obtained
+	// for a long time on the server side.
+	client := &http.Client{Timeout: 120 * time.Second}
```
Timeout duration needs to be in a const.

lbn commented on main.go:
```diff
@@ -0,0 +1,182 @@
+	createOptions["skip_ui"] = true
+
+	// Setup revizor container branches
+	branch := os.Getenv("BRANCH")
```
These could be parsed in a separate function (e.g. newCreateOptions) and then passed as params to `doCreate` to keep the logic in this function focused on creating the container rather than parsing envvars.

lbn commented on main.go:
```diff
@@ -0,0 +1,182 @@
+
+func doCreate() error {
+	log.Println("Creating the container...")
+	createOptions := make(map[string]interface{})
```
We know the structure of the Revizor API request so we can define it as a struct.

lbn commented on main.go:
```diff
@@ -0,0 +1,182 @@
+	client := &http.Client{Timeout: 120 * time.Second}
+	resp, err := client.Do(req)
+	if err != nil {
+		log.Fatal(err)
```
Would be nice to add some context to errors like this like for the StatusCode != 201 error.

lbn commented on main.go:
```diff
@@ -0,0 +1,182 @@
+	if resp.StatusCode != 201 {
+		log.Fatalf("Cannot create the container: %d", resp.StatusCode)
+	}
+	defer resp.Body.Close()
```
Make sure to check if `resp.Body` is nil first and exit if it is.

lbn commented on main.go:
```diff
@@ -0,0 +1,182 @@
+	// set-output: GitHub Action mechanism that sets the output parameter.
+	fmt.Printf("::set-output name=container_id::%s\n", cont.ID)
+	fmt.Printf("::set-output name=hostname::%s.%s\n", cont.ID, teBaseURL)
+	for i := 1; i <= 10; i++ {
```
Need to create consts for max retries and delay.

lbn commented on main.go:
```diff
@@ -0,0 +1,182 @@
+	}
+	log.Printf("The container %s has been created", cont.ID)
+	// set-output: GitHub Action mechanism that sets the output parameter.
+	fmt.Printf("::set-output name=container_id::%s\n", cont.ID)
```
Would be great to have a separate function to set outputs.

lbn commented on main.go:
```diff
@@ -0,0 +1,182 @@
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil || resp.StatusCode != 202 {
+		return fmt.Errorf("Failed to delete the container, status code %d", resp.StatusCode)
```
`resp` could be nil here. Need to check for err and return it in a separate if statement before operating on `resp`.

lbn commented on main.go:
```diff
@@ -0,0 +1,182 @@
+	client := &http.Client{}
+	resp, err := client.Do(req)
+	if err != nil || resp.StatusCode != 200 {
+		return fmt.Errorf("The healthcheck error: %s", resp.Status)
```
If err is not nil, resp could be nil so need to check err and return in a separate if statement.

lbn commented on main.go:
```diff
@@ -0,0 +1,182 @@
+		log.Fatal(err)
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", scalrToken))
+	req.Header.Set("Prefer", "profile=preview")
```
Would be great to have in a const as well.

vmotso replied:
We will change it after [SCALRCORE-17546](https://scalr-labs.atlassian.net/browse/SCALRCORE-17546).

vmotso committed:
Use the scratch image for runtime

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso committed:
Make requested changes

Use Revizor API to create test containers to run acceptance tests against
SCALRCORE-16917

vmotso commented:
After requested changes, everything works: https://github.com/vmotso/check-gh-action-revizor/pull/1/checks
